lib/spack/spack/compilers/msvc.py

```diff
@@ -29,6 +29,90 @@
 }
 
 
+class CmdCall:
+    """Compose a call to `cmd` for an ordered series of cmd commands/scripts"""
+
+    def __init__(self, *cmds):
+        if not cmds:
+            raise RuntimeError(
+                """Attempting to run commands from CMD without specifying commands.
+                Please add commands to be run."""
+            )
+        self._cmds = cmds
+
+    def __call__(self):
+        out = subprocess.check_output(self.cmd_line, stderr=subprocess.STDOUT)  # novermin
+        return out.decode("utf-16le", errors="replace")  # novermin
+
+    @property
+    def cmd_line(self):
+        base_call = "cmd /u /c "
+        commands = " && ".join([x.command_str() for x in self._cmds])
+        # If multiple commands are being invoked by a single subshell
+        # they must be encapsulated by a double quote. Always double
+        # quote to be sure of proper handling
+        # cmd will properly resolve nested double quotes as needed
+        #
+        # `set`` writes out the active env to the subshell stdout,
+        # and in this context we are always trying to obtain env
+        # state so it should always be appended
+        return base_call + f'"{commands} && set"'
+
+
+class VarsInvocation:
+    def __init__(self, script):
+        self._script = script
+
+    def command_str(self):
+        return f'"{self._script}"'
+
+    @property
+    def script(self):
+        return self._script
+
+
+class VCVarsInvocation(VarsInvocation):
+    def __init__(self, script, arch, msvc_version):
+        super(VCVarsInvocation, self).__init__(script)
+        self._arch = arch
+        self._msvc_version = msvc_version
+
+    @property
+    def sdk_ver(self):
+        """Accessor for Windows SDK version property
+
+        Note: This property may not be set by
+        the calling context and as such this property will
+        return an empty string
+
+        This property will ONLY be set if the SDK package
+        is a dependency somewhere in the Spack DAG of the package
+        for which we are constructing an MSVC compiler env.
+        Otherwise this property should be unset to allow the VCVARS
+        script to use its internal heuristics to determine appropriate
+        SDK version
```

Comment:
(question, no change needed) This is true if we are building `win-sdk` itself. And also for DAGs that have no dependency on `win-sdk`.

In the latter case, I'm unsure how the following applies:

> Otherwise this property should be unset to allow the VCVARS script to use its internal heuristics to determine appropriate SDK version

It sounds like all MSVC builds have some dependency on `win-sdk` (though right now in spack it is not possible to record this).

To be clear, I'm not worried about link consistency: DAGs with `win-sdk` anywhere in their dependencies will invoke this consistently given the way you coded it. Rather, I'm just trying to keep track of places where modeling constraints lead to confusing outcomes. If `msvc` itself doesn't need `win-sdk` in all cases, that is worth knowing too.

Reply:
> It sounds like all MSVC builds have some dependency on win-sdk (though right now in spack it is not possible to record this).
Yes and no. No, not all MSVC builds include/link to components of the Windows SDK, however, the VCVARS script will still load SDK variables into the development env regardless of whether or not they're used, and not setting them appropriately can result in issue in some circumstances. So long as the VCVARS sets a consistent version (it detects its own, or you give it a specific one) generally all is well.

Reply:
> can result in issue in some circumstances.

it would be interesting to know what those are, but that question need not hold up this review.